pkg/repo/instanceid.go

```diff
@@ -0,0 +1,99 @@
+package repo
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/denisbrodbeck/machineid"
+)
+
+// GenerateInstanceID creates a unique, anonymous identifier for the instance of bacalhau.
+// It combines the machine ID and MAC address to ensure uniqueness across different
+// environments, including virtual machines and cloud instances.
```

Comment:
Maybe worth clarifying in the comment:
We only attempt to generate an instance id if no id was found the `system_metadata`, which can happen during new node initialization or if the DataDir was wiped. We attempt to generate deterministic id for analytics purposes and track the usage of the instance.
 
-- 

Now since it is best effort, rather than failing because of errors when trying to extract the machine id or mac address, why don't we fallback to `hash(uuid)`? This will be sufficient 99.999% of the time since we do persist it in the system_metadata

My other question is, if the user is running a node or a client against different networks with different `DataDir`, we will assign them the same InstanceID. Maybe this is the desired behaviour, but worth thinking about

Reply:
There are several clarifying comments related to when we generate the instanceID, and how we persist it in the repo Open and Init methods. I think usage is well covered there. (It should be easier to grok now that I've fixed the mixed commits :grinning:)

> Now since it is best effort, rather than failing because of errors when trying to extract the machine id or mac address, why don't we fallback to hash(uuid)?

What value would including this random string (UUID) add? 
The purpose of the instanceID is to identify (anomalously) unique instances of bacalhau. I'd much prefer an empty value - which implies a value couldn't be derived - than a random string with no relationship to the actual instance. My preference is to leave the field empty if it couldn't be derived.

> This will be sufficient 99.999% of the time since we do persist it in the system_metadata

I've made a small change since your last review: We now generate and persist an instanceID when the repo is first initialized and **also each time the repo is opened**. This ensures the instanceID is associated with the machine running bacalhau, even if the repo is copied to a different machine. In this sense we are treating the `system_metadata.yaml` file a bit like a cache. In work to follow this, we can read the instanceID from the `system_metadata.yaml` file and include it in the job metadata and/or HTTP headers when an instance interacts with the network.

> My other question is, if the user is running a node or a client against different networks with different DataDir, we will assign them the same InstanceID. Maybe this is the desired behaviour, but worth thinking about

Right, I think this behavior makes sense. The same instance of bacalhau is free to interact with many networks. A measure of this interaction could be derived be counting the number of different networks an instance has interactions with.

Reply:
> What value would including this random string (UUID) add?
> The purpose of the instanceID is to identify (anomalously) unique instances of bacalhau. I'd much prefer an empty value - which implies a value couldn't be derived - than a random string with no relationship to the actual instance. My preference is to leave the field empty if it couldn't be derived.

How a _persisted_ UUID doesn't provide a relationship to the instance? Our goal is to identify the instance the jobs are submitted from (client) or executing the job (node). Assigning and persisting a UUID when you first initialize the client/node is more than enough and fulfils this purpose. What you are trying to do here with generating idempotent IDs is just to handle the edge cases where the repo is deleted and we still want to tag jobs with the same instance, which is great, but lets treat it as best effort and fallback to UUID instead of not tracking at all

Keep in mind this means we only attempt to generate an ID if no ID was found in the system_metadata.yaml, and don't attempt to override the value

Reply:
Can you address these comments so we can close the PR?

Reply:
> this means we only attempt to generate an ID if no ID was found in the system_metadata.yaml, and don't attempt to override the value

Sounds good, will address.

Reply:
replaced by https://github.com/bacalhau-project/bacalhau/pull/4411